updates/build-log/memo/monitoring.md

```diff
@@ -0,0 +1,247 @@
+  arguments: {
+    username: "Memo NFT",
+    webhookUrl: process.env.DISCORD_WEBHOOK_URL,
+    content: generatedMessage,
+    title: "📢 New notes minted in the Memo",
+    autoFormat: true,
+  }
+});
+```
+
+This system ensures both optimal performance and high reliability through:
+
+* **Connection Resilience**: Implements a retry mechanism for establishing the initial connection.
+* **Exponential Backoff**: If a notification fails, the system waits progressively longer before retrying, preventing API rate-limiting issues.
+* **Graceful Error Handling**: Failures are logged without crashing the parent workflow.
+
+#### CI/CD workflow monitoring
+
+We have integrated instrumental monitoring directly into our **GitHub Actions** workflows. This provides real-time status updates on deployments and other critical automated processes, sending success or failure notifications to a dedicated Discord channel.
+
+**Standardized Notification Steps:**
+
+```yaml
+- name: Notify Discord on Success
+  if: success()
+  uses: sarisia/actions-status-discord@v1
+  with:
+    webhook: ${{ secrets.DISCORD_WEBHOOK_URL }}
+    title: '✅ Deployment Completed Successfully'
+    color: 0x00ff00
+
+- name: Notify Discord on Failure
+  if: failure()
+  uses: sarisia/actions-status-discord@v1
+  with:
+    webhook: ${{ secrets.DISCORD_WEBHOOK_URL }}
+    title: '❌ Deployment Failed'
+    description: 'Please check workflow logs for details.'
+    color: 0xff0000
+```
+
+### System architecture and implementation
+
+Our monitoring platform is built on a modern, flexible technology stack chosen for its performance, ease of integration, and alignment with our team's expertise.
+
+#### The DuckDB analytics engine
+
+We selected **DuckDB** as our core data processing engine due to its unique capabilities for in-process analytics. It allows us to perform complex analytical queries on local and remote data sources without the overhead of a traditional data warehouse.
+
+**Key Advantages:**
+
+* **Federated Queries**: Natively queries remote Parquet files and connects to our live **PostgreSQL** database in a single session.
+* **High Performance**: Columnar-vectorized query execution is highly optimized for analytical workloads.
+* **Zero-Dependency**: Runs entirely in-memory, simplifying our CI/CD environment setup.
+
+**PostgreSQL Integration:**
+The connection process involves loading the `postgres` extension and attaching our remote database as a read-only source.
+
+```typescript
+async function setupDuckDBConnections(): Promise<DuckDBConnection> {
+  const instance = await DuckDBInstance.create(':memory:');
+  const connection = await instance.connect();
+  
+  // Load extensions and attach remote database
+  await connection.runAndReadAll('LOAD postgres;');
+  await connection.runAndReadAll(`
+    ATTACH '${process.env.DB_CONNECTION_STRING}' AS memo_nft_db (TYPE postgres, READ_ONLY);
+  `);
+
+  return connection;
+}
+```
+
+### Security and compliance framework
```

Comment:
Phần này chắc humble đi tý. Kiểu apply security best practices thôi ông. với xem thử wording lại phần example truncate wallet bên dưới sound trivial quá. rồi dedicate cả phần "Compliance and privacy" nói mỗi cái truncate hơi căng. ông xem gộp mấy section bên chỗ security lại hết rồi mention ngắn gọn cái kia làm vd thôi